### DISCUSSION TITLE
[Question] Control Refs AttributeError for Row and Column if ".current.controls." is not in a "def XXX(e)"

### DISCUSSION THREAD

**Discussion by heximing** [flet-dev/flet#1519] [Q&A] [answered]
### Question

Hi I am new using flet (from tkinter). I realize, when using Control Refs, "ft.Row" and "ft.Column" cannot use ".current.controls." like other controls like TextField if I use it in main (not in another function).

in the official Introduction (https://flet.dev/docs/guides/python/getting-started), the sample code is:

```
import flet as ft

def main(page):

    first_name = ft.Ref[ft.TextField]()
    last_name = ft.Ref[ft.TextField]()
    greetings = ft.Ref[ft.Column]()

    def btn_click(e):
        greetings.current.controls.append(
            ft.Text(f"Hello, {first_name.current.value} {last_name.current.value}!")
        )
        first_name.current.value = ""
        last_name.current.value = ""
        page.update()
        first_name.current.focus()

    page.add(
        ft.TextField(ref=first_name, label="First name", autofocus=True),
        ft.TextField(ref=last_name, label="Last name"),
        ft.ElevatedButton("Say hello!", on_click=btn_click),
        ft.Column(ref=greetings),
    )

ft.app(target=main)
```

The "greetings" are modified in "btn_click(e)".
However, if it is not in a function, it would throw an error:
`AttributeError: 'NoneType' object has no attribute 'controls'`
I supposed this is intentional, but how could I use Row or Column as Refs without a separate function?

### Code sample

```python
import flet as ft


def main(page):

    first_name = ft.Ref[ft.TextField]()
    last_name = ft.Ref[ft.TextField]()
    greetings = ft.Ref[ft.Column]()

    greetings.current.controls.append(
        ft.Text(f"whatever text here 111"),
        ft.Text(f"whatever text here 222"),
        ft.Text(f"whatever text here 333"),
    )

    page.add(
        ft.TextField(ref=first_name, label="First name", autofocus=True),
        ft.TextField(ref=last_name, label="Last name"),
        ft.Column(ref=greetings),
    )

ft.app(target=main)
```

### Error message

```bash
AttributeError: 'NoneType' object has no attribute 'controls'
```

### ------------------------------------------------------

- [X] I have searched for answers to my question both in the issues and in previous discussions.

**Comment by ndonkoHenri** [comment, marked as answer]
You need to place `greetings.control....` after the `page.add(...)` for it to work.
The refs first need to be registered (as in your `ref=greetings`) before used. Before this 'binding' the `ref.current` is `None` since it is not connected to a control.

**Reply by heximing** [reply]
Thanks very much. That totally makes sense.
